File: LeetCode/0328/golang/lc0328.go
package lc0328

func oddEvenList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	headEven := &ListNode{Val: -1, Next: nil}
	headOdd := &ListNode{Val: -1, Next: nil}

	currEven := headEven
	currOdd := headOdd
	isEven := false

	for head != nil {
		if isEven {
			currEven.Next = head
			currEven = head
		} else {
			currOdd.Next = head
			currOdd = head
		}

		head = head.Next
		isEven = !isEven
	}

	// Reconstruct the list: first all the odd nodes, then all the even nodes.
	currOdd.Next = headEven.Next
	currEven.Next = nil
	return headOdd.Next
}

File: LeetCode/0328/python/lc0328.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Runtime complexity: O(n)
# Auxiliary space complexity: O(1)
# Subjective level: medium-
# Solved on: 2026-02-17
class Solution:
    def oddEvenList(self, head: ListNode | None) -> ListNode | None:
        if head is None or head.next is None:
            return head

        headOdd, headEven = head, head.next
        currOdd, currEven = headOdd, headEven
        head = head.next.next
        isOdd = True # Linked list elements are 1-indexed, and we're currently at head+2 (index 3).
        while head is not None:
            if isOdd:
                currOdd.next = head
                currOdd = head
            else:
                currEven.next = head
                currEven = head
            head = head.next
            isOdd = not isOdd

        # Finally link the even nodes after all the odd nodes:
        currOdd.next = headEven
        currEven.next = None
        return headOdd
